### ISSUE TITLE
Some Course Suggestions

### ISSUE THREAD

**Issue by SandyThompson** [githubteacher/flaming-octo-meme#138]
Consider restructuring this course by standard workflows.  I can already tell that the first thing I'm going to do when I start using git is to hit up Google for a step-by-step workflow description.
1. how to edit files in a repository
   - create a branch
   - edit the file
   - put it into the staging area
   - ... etc.
2. how to add files to a repository
   - ... etc.
3. how to delete files from a repository
4. how to merge conflicting changes
5. how to view the history of all the changes to a file/folder
6. how to rollback a change (both most recent and not)
7. ... etc.

Some other basic suggestions:
1. favour a gui over the command line
2. spend the first 15 minutes talking about the git technology relative to other major systems (pros and cons)
3. provide a glossary of terms and their counterparts in other major systems
   - fork
   - branch
   - pull request
   - push
   - ... etc

**Comment by githubstudent** [comment]
@SandyThompson did you attend yesterday's class? Many of your questions about workflow were answered there. These classes are really meant to flow together and not be independent!

**Comment by SandyThompson** [comment]
Thanks for the response.

I did attend yesterday, as did a number of other people here in the room with me.  We're making the same sorts of comments during breaks, and I thought it might be useful if I posted my version of these sentiments here.  Perhaps there is a mindset shift that needed to happen prior to attending the first class, but I'm finding myself thoroughly confused and constantly struggling to catch-up with what's going on and how it relates to my daily tasks of check-out / update / check-in work.

If there are announcements being made such as "here is an end-to-end file change workflow", I completely missed them.  The topics I'm referring to _are_ being covered, but are not clearly distinct from a larger conversation which includes commentary about what appear to be uncommonly-used features, command line options, trouble-shooting unexpected issues, etc.  The sorting out of the essential from the extraneous is something that's happening in my head after a topic is complete, and while the next one has already started.

**Comment by DiegoALeon** [comment]
Yeah. I've looking at youtube videos too to clarify a number of doubts.
Going over the graphical user interface in the course would be a good thing, since the stuff that the command line prompt spits out is not always intelligible to different types of users. Especially when there are conflicts, etc.

**Comment by bmatiu** [comment]
I don't get if there is a paradigm shift with Git or it is simply overly-complicated.

In any previous source control systems:
One time work space creation.

A regular workflow with traditional source controls systems:
 Work on any files that are automatically checked out. 
 Create a changelist. 
 Check in. (3 steps)

The way I understand it now, with git there are more steps (am I missing something?)
Create branch 
Checkout branch
Create issue.
Work on files.
Commit.
Push
Pull request.  (7 steps?)

I don't get when an action is finished? When something gets closed? When I need to do what?

In the end it is supposed to upload and download from a server? And allow some comments and some history.

**Comment by githubstudent** [comment]
@SandyThompson & @DiegoALeon I do appreciate your feedback! What I am hearing is that we need to spend more high-level time on the workflow itself. I will make a note for patrick to do this during the Q&A session. In the meantime, we are covering things that are more file level operations so hopefully this will be easier to follow without this context!

**Comment by githubstudent** [comment]
@bmatiu the workflow looks something like this:
- `git checkout -b <branch-name>` <-- creates your branch and checks it out
- work on file
- `git add <filename>` <-- puts the file in staging
- `git commit` <-- commits your changes
- `git push -u origin <branch-name>` <--- sends your changes to the remote
- create pull request
- merge pull request into master!

The paradigm shift is that this is distributed version control vs. central version control. With DVCS you are not constantly connected to the remote. You explicitly tell git when you want to push and pull changes. This gives you greater control over your workflow while still promoting the collaboration features of github!

**Comment by bmatiu** [comment]
Thank you for the workflow!

Regarding the statement that with central version control you always have to be connected I do not agree. This is why reconcile offline work exists. 

Regarding the fact that you push and pull (in Git terms) to an from server, you check out and check in from and to the server (in older language). So why there is a need for more steps? What do we gain by this?

**Comment by SandyThompson** [comment]
Thanks @githubstudent.  I suspect if I had workflow bullet lists like that for the 5 or 10 most commonly performed actions and the most commonly encountered exceptions to the standard workflow, I'd be all set with no need to attend a course!  :-)

**Comment by githubstudent** [comment]
@SandyThompson we would miss you if you didn't attend the course! BUT, we do try to put as much information as possible online for people who prefer to learn independently. I think [guides.github.com](https://guides.github.com/) is a great place to start!
